src/System.IO/tests/MemoryStream/MemoryStreamTests.cs

```diff
@@ -0,0 +1,219 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
+using System.Threading.Tasks;
+using System.Threading;
+
+namespace System.IO.Tests
+{
+    public class MemoryStreamTests
+    {
+        #region Test Methods
+
+        [Fact]
+        public static void WriteBeyondEndTest()
+        {
+            MemoryStream memoryStream = new MemoryStream();
+
+            long origLength = memoryStream.Length;
+            byte[] bytes = new byte[10];
+            for (int i = 0; i < bytes.Length; i++)
+                bytes[i] = (byte)i;
+            int spanPastEnd = 5;
+            memoryStream.Seek(spanPastEnd, SeekOrigin.End);
+            Assert.Equal(memoryStream.Length + spanPastEnd, memoryStream.Position);
+
+            // Test Write
+            memoryStream.Write(bytes, 0, bytes.Length);
+            long pos = memoryStream.Position;
+            Assert.Equal(pos, origLength + spanPastEnd + bytes.Length);
+            Assert.Equal(memoryStream.Length, origLength + spanPastEnd + bytes.Length);
+            WritePastEndHelper(memoryStream, bytes, origLength, spanPastEnd, false);
+
+            // Test WriteByte
+            origLength = memoryStream.Length;
+            memoryStream.Position = memoryStream.Length + spanPastEnd;
+            memoryStream.WriteByte(0x42);
+            long expected = origLength + spanPastEnd + 1;
+            Assert.Equal(expected, memoryStream.Position);
+            Assert.Equal(expected, memoryStream.Length);
+
+            // Test WriteAsync
+            origLength = memoryStream.Length;
+            memoryStream.Position = memoryStream.Length + spanPastEnd;
+            Task writeTask = memoryStream.WriteAsync(bytes, 0, bytes.Length, new CancellationToken());
+            writeTask.Wait();
```

Comment:
Instead of the .Wait(), I'd suggest doing:
```C#
Assert.Equal(TaskStatus.RanToCompletion, writeTask.Status);
```
WriteAsync on MemoryStream should complete synchronously.